### PR TITLE
Ale integration into dev 

### PR DESCRIPTION
**Pull request by Kelvinrr**
## Description

 
Pushing Ale integration branch into dev branch. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [ ] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/base/objs/Blob/Blob.cpp [hunk 1]
```diff
@@ -150,6 +150,43 @@ namespace Isis {
     return p_blobName;
   }
 
+  void Blob::ReadString(const QString &write_string){
+    // Read the binary data
```
Comment (acpaquette) [inline isis/src/base/objs/Blob/Blob.cpp]: Could use a doc string.

File: isis/src/base/objs/Blob/Blob.cpp [hunk 2]
```diff
@@ -150,6 +150,43 @@ namespace Isis {
...
+                   " [" + p_blobName + "]";
+      throw IException(IException::Io, msg, _FILEINFO_);
+    }
+
+    stream.read(p_buffer, p_nbytes);
+    if (!stream.good()) {
+      QString msg = "Error reading data from " + p_type + " [" + p_blobName + "]";
+      throw IException(IException::Io, msg, _FILEINFO_);
+    }
+  }
```
Comment (acpaquette) [inline isis/src/base/objs/Blob/Blob.cpp]: Any particular reason this got shuffled around?

File: isis/src/base/objs/PvlObject/PvlObject.cpp [hunk 3]
```diff
@@ -59,6 +59,49 @@ namespace Isis {
...
 
 
+ 
+  /**
+   * Creates a PvlObject with a specified name converted 
+   * from a json object.
+   *
+   * TODO: Needs to be more robust, right now exists in order to 
+   * convert an Ale ISD NaifKeywords group over to an ISIS 
+   * compatible version.
```
Comment (acpaquette) [inline isis/src/base/objs/PvlObject/PvlObject.cpp]: Was this handled or is there more that needs to happen here?

**Review by acpaquette — COMMENTED** [review]
A few small things, overall this looks good!

File: isis/src/base/objs/PvlObject/PvlObject.cpp (hunk 3 above)
Comment (Kelvinrr) [reply isis/src/base/objs/PvlObject/PvlObject.cpp]: It seems to work so far, simply doesn't handle all types. Probably something that should change in the near future. Not sure if super important right now.

File: isis/src/base/objs/Blob/Blob.cpp (hunk 2 above)
Comment (Kelvinrr) [reply isis/src/base/objs/Blob/Blob.cpp]: Not entirely sure but really shouldn't matter, if it's an issue I can commit and quick reshuffle.

File: isis/cmake/FindAle.cmake [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+find_library(ALE_LIBRARY
+  NAMES ale
+)
+
+message(STATUS "ALE INCLUDE DIR: "  ${ALE_INCLUDE_DIR} )
+message(STATUS "ALE LIB: "  ${ALE_LIBRARY} )
+
+get_filename_component(ALE_ROOT_INCLUDE_DIR "${ALE_INCLUDE_DIR}" DIRECTORY)
```
Comment (jessemapel) [inline isis/cmake/FindAle.cmake:19]: This is a note for ALE, we should generate a FindALE script as part of ALE.

File: isis/src/base/objs/Blob/Blob.cpp (hunk 1 above)
Comment (jessemapel) [reply isis/src/base/objs/Blob/Blob.cpp]: Also needs a test.

File: isis/src/base/objs/PvlObject/PvlObject.cpp (hunk 3 above)
Comment (jessemapel) [reply isis/src/base/objs/PvlObject/PvlObject.cpp]: I think this is one of those things were we can expand the functionality as needed. I would remove this TODO and instead document which types it converts right now.

File: isis/src/base/objs/SpicePosition/SpicePosition.cpp [hunk 5]
```diff
@@ -337,6 +337,39 @@ namespace Isis {
   }
 
 
+  void SpicePosition::LoadCache(json &isdPos) {
```
Comment (jessemapel) [inline isis/src/base/objs/SpicePosition/SpicePosition.cpp:351]: Needs a doc string

File: isis/src/base/objs/SpiceRotation/SpiceRotation.cpp [hunk 6]
```diff
@@ -402,6 +404,66 @@ namespace Isis {
   }
 
 
+  void SpiceRotation::LoadCache(json &isdRot){
```
Comment (jessemapel) [inline isis/src/base/objs/SpiceRotation/SpiceRotation.cpp:418]: Needs a doc string

File: isis/src/base/objs/SpiceRotation/SpiceRotation.cpp [hunk 7]
```diff
@@ -402,6 +404,66 @@ namespace Isis {
...
+
+    }
+    else {
+      p_TC.resize(9);
+      ident_c((SpiceDouble( *)[3]) &p_TC[0]);
+    }
+
+    p_source = Memcache;
+    SetEphemerisTime(p_cacheTime[0]);
+  }
```
Comment (jessemapel) [inline isis/src/base/objs/SpiceRotation/SpiceRotation.cpp:459]: Needs a test

File: isis/src/base/objs/SpicePosition/SpicePosition.cpp [hunk 8]
```diff
@@ -337,6 +337,39 @@ namespace Isis {
...
+        std::vector<double> vel = {it->at(0).get<double>(), it->at(1).get<double>(), it->at(2).get<double>()};
+        p_cacheVelocity.push_back(vel);
+      }
+    }
+
+    p_hasVelocity = !p_cacheVelocity.empty();
+    p_source = Memcache;
+
+    SetEphemerisTime(p_cacheTime[0]);
+  }
```
Comment (jessemapel) [inline isis/src/base/objs/SpicePosition/SpicePosition.cpp:382]: Needs a test

**Comment by jessemapel** [comment]
I am going to work on a PR to fix the docs and test issues now.

**Review by jessemapel — APPROVED** [review]
This is now ready to go. I am going to merge it later today unless anyone has more to add.

**Comment by Kelvinrr** [comment]
Gonna go ahead and merge so we can use the changes in the Rotation/Position refactors

**Pull request merged**